tools/build-bpf.py

```diff
@@ -0,0 +1,169 @@
+        if not hexdump_buffer:
+            hexdump_buffer = get_default_hexdump_buffer_name(out_file)
+
+        out_file_fd.write(hexdump_src_template
+                .replace("%src_c%", src_c)
+                .replace("%buffer_name%", hexdump_buffer)
+                .replace("%hexdump%", hexdump(in_file, indent=6)))
+
+def bpf_build(args):
+    clang_flags = [
+            '-Wno-compare-distinct-pointer-types',
+            '-O2',
+            '-target',
+            'bpf',
+            '-emit-llvm',
+            '-c',
+    ]
+    with tempfile.NamedTemporaryFile(mode='rb', suffix='.bc') as clang_out_fd, \
+        tempfile.NamedTemporaryFile(mode='rb', suffix='.o') as llc_out_fd:
+        clang_compile(clang_exec=args.clang_exec,
+                clang_flags=clang_flags, src_c=args.bpf_src_c,
+                src_h=args.bpf_src_h, out_file=clang_out_fd.name,
+                architecture=get_architecture_compiler_flag(args.arch),
+                glibc_headers=args.glibc_headers);
+
+        llc_flags = [
+            '-march=bpf',
+            '-filetype=obj',
+            '-o'
+        ]
+        llc_compile(llc_exec=args.llc_exec, llc_flags=llc_flags,
+        in_file=clang_out_fd.name, out_file=llc_out_fd.name)
+
+        gen_hexdump_header_file(in_file=llc_out_fd.name,
+                out_file=args.bpf_hexdump_h,
+                src_c=args.bpf_src_c,
+                hexdump_buffer=args.bpf_hexdump_buffer)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+            '--clang_exec',
+            type=str,
+            help='Path to clang exec.')
+
+    parser.add_argument(
+            '--llc_exec',
+            type=str,
+            help='Path to llc exec.')
+
+    parser.add_argument(
+            '--bpf_src_h',
+            type=str,
+            nargs='*',
+            help='Path to *.h header of compiled BPF program in systemd tree.')
+
+    parser.add_argument(
+            'bpf_src_c',
+            type=str,
+            help='Path to *.c source of BPF program in systemd source tree. \
+                    Relative to invoke directory.')
+
+    parser.add_argument(
+            'bpf_hexdump_h',
+            type=str,
+            help='*.h filename to write byte array representation (hexdump) \
+                    of BPF *.o object file.')
+
+    parser.add_argument(
+            '--bpf_hexdump_buffer',
+            type=str,
+            help='Name of unsigned char[] C array storing hexdump. If not \
+                    specified, defaults to basename of bpf_hexdump_h plus \
+                    `_buffer` suffix, e.g. `hexdump_buffer` for \
+                    `hexdump.h`.')
+
+    parser.add_argument(
+        '--glibc_headers',
+        type=str,
+        help='Include directory of glibc header files. Default \
+                `/usr/include/x86_64-linux-gnu/` for x86_64 Debian.',
+        default='/usr/include/x86_64-linux-gnu/')
```

Comment:
We shouldn't have any silent defaults here. This should be queried from inside meson (so it honors pkgconfig) and passed down to the python script.

Reply:
@flokli 

Removed that band-aid, added a check for [target triplet](https://wiki.osdev.org/Target_Triplet) taken from `gcc -dumpmachine` output.

The rationale for addition search path is due to on Debian-like distribution, the default clang search paths for `-target bpf` doesn’t include multi architecture paths, e.g.  `/usr/include/x86_64-linux-gnu/`. That complicates inclusion of `<linux/types.h>` header into BPF programs because of `<asm/types.h>` inclusion which was moved to multiarch path from `/usr/include`. Please see [[0]](https://www.spinics.net/lists/xdp-newbies/msg01770.html) [[1]](https://patchwork.ozlabs.org/project/netdev/patch/20200311123421.3634-1-tklauser@distanz.ch/) for further description of the issue.
Another solution would be adding `gcc-multilb` package with `/usr/include/asm/types.h` to e.g. TravisCI, but given the size of the package, doesn't worth is IMO.

I'm not a meson expert but it seems that it's not that flexible, e.g. `meson.get_compiler('c')` implies `gcc` and it is not possible to choose between several compilers and get a set of compiler helpers for `clang` instead.  
I think in the EoD `bpf_object` rule should be natively supported by meson as opposed to custom rule with build script. Then all low-level tweaking, including compiler search paths, can be done by meson under the hood.

[0] https://www.spinics.net/lists/xdp-newbies/msg01770.html
[1] https://patchwork.ozlabs.org/project/netdev/patch/20200311123421.3634-1-tklauser@distanz.ch/